sbin/tpm2-evt-log-parser.awk

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/gawk -bf
+@load "readfile"
+
+function assert(condition, string)
+{
+	if (!condition) {
+		print string
+		exit 1
+	}
+}
+
+function ord_init()
+{
+	for (_i = 0; _i < 256; _i++) {
+		ord[sprintf("%c", _i)] = _i
+	}
+}
+
+function x2n(hex, width)
+{
+	mult = 1
+	num = 0
+	for (_i = 0; _i < width; _i++) {
+		num += ord[substr(hex, _i+1, 1)] * mult
+		mult *= 256
+	}
+	return num
+}
+
+function hexdump(hex, len)
+{
+	for (_i = 0; _i < len; _i++) {
+		printf("%02x", ord[substr(hex, _i+1, 1)])
+	}
+}
+
+function alg_name(id)
+{
+	switch (id) {
+		case 0x0004: return "SHA1"
+		case 0x000b: return "SHA256"
+		case 0x000c: return "SHA384"
+		case 0x000d: return "SHA512"
+		case 0x0012: return "SM3-256"
+		case 0x0027: return "SHA3-256"
+		case 0x0028: return "SHA3-384"
+		case 0x0029: return "SHA3-512"
+		default: return sprintf("unknown (%#06)", id)
+	}
+}
+
+function string_or_hex(str, len)
+{
+	_len = len
+	if (_len > 300)
+		_len = 128
+	# String must start with a series of printable characters ...
+	if (match(str, "[[:graph:][:blank:]]*", a) != 1) {
+		hexdump(str, _len)
+	# ... long until the end, with "optional" (i.e. bad implementation) \0.
+	} else if (len != a[0, "length"] &&
+		   (len != a[0, "length"] + 1 || index(str, "\0") != len)) {
```

Comment:
How about doing this (like in `cbmem`):<details><summary>patch</summary>
<p>

```diff
 sbin/tpm2-evt-log-parser.awk | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/sbin/tpm2-evt-log-parser.awk b/sbin/tpm2-evt-log-parser.awk
index 00fab7a..5a66708 100755
--- a/sbin/tpm2-evt-log-parser.awk
+++ b/sbin/tpm2-evt-log-parser.awk
@@ -49,6 +49,21 @@ function alg_name(id)
 	}
 }
 
+function is_zero_padded(str, len)
+{
+	for (_i = 0; _i < len; _i++) {
+		if (ord[substr(str, _i+1, 1)] == 0) {
+			break
+		}
+	}
+	for (; _i < len; _i++) {
+		if (ord[substr(str, _i+1, 1)] != 0) {
+			return 0
+		}
+	}
+	return 1
+}
+
 function string_or_hex(str, len)
 {
 	_len = len
@@ -58,8 +73,7 @@ function string_or_hex(str, len)
 	if (match(str, "[[:graph:][:blank:]]*", a) != 1) {
 		hexdump(str, _len)
 	# ... long until the end, with "optional" (i.e. bad implementation) \0.
-	} else if (len != a[0, "length"] &&
-		   (len != a[0, "length"] + 1 || index(str, "\0") != len)) {
+	} else if (len != a[0, "length"] && !is_zero_padded(str, _len)) {
 		hexdump(str, _len)
 	} else
 		printf("%.*s", _len, a[0])
```

</p>
</details> 

Same for the other awk file.  Without this change coreboot's log is printed as a hex dump.

Reply:
I thought about it, but this treats some of TXT fields as strings when they really aren't, e.g. `EVTYPE_OSSINITDATA_
CAP_HASH` value of `0x00000026` (for TPM 1.2) or entries with all 0s.